Sources/SourceControl/RepositoryManager.swift

```diff
@@ -165,54 +121,74 @@ public class RepositoryManager {
     /// Get a handle to a repository.
     ///
     /// This will initiate a clone of the repository automatically, if
-    /// necessary, and immediately return. The client can add observers to the
-    /// result in order to know when the repository is available.
-    public func lookup(repository: RepositorySpecifier) -> RepositoryHandle {
-        // Check to see if the repository has been provided.
-        if let handle = repositories[repository.url] {
-            // Fetch and update the repository when it is being looked up.
-            // We need some mechanism to maybe not do this for pinned repos.
+    /// necessary.
+    public func lookup(repository: RepositorySpecifier, completion: @escaping (Result<RepositoryHandle, AnyError>) -> Void) {
+        DispatchQueue.global().async {
+            repostoriesLock.whileLocked {
+                // Check to see if the repository has been provided.
+                if let handle = self.repositories[repository.url] {
+                    // Fetch and update the repository when it is being looked up.
+                    // We need some mechanism to maybe not do this for pinned repos.
+                    do {
+                        let repo = try handle.open()
+                        try repo.fetch()
+                    } catch {
+                        // FIXME: Better error handling.
+                        // Maybe just print a warning that we couldn't fetch it
+                        // since we already have the repo.
+                        print("Error while fetching the repo: \(handle)")
+                    }
+                    completion(Result(handle))
+                } else {
+                    // Otherwise, fetch the repository and return a handle.
+                    let subpath = RelativePath(repository.fileSystemIdentifier)
+                    let handle = RepositoryHandle(manager: self, repository: repository, subpath: subpath)
+                    self.repositories[repository.url] = handle
+                }
+            }
+
+            // Ensure nothing else exists at the subpath.
+            let repositoryPath = self.path.appending(subpath)
+            if localFileSystem.exists(repositoryPath) {
+                _ = try? removeFileTree(repositoryPath)
+            }
+
+            let result: Result<RepositoryHandle, AnyError>
```

Comment:
This could use `Result.init(_ body: () throws -> Value)` to be simpler, I think.